File: backend/app/domain/rules.py
"""Configurable match-rules model — CricNetra's headline differentiator.

The rulebook is *data*, not code: an organizer composes a `MatchRules` object
(directly or from a preset), names it, and reuses it. The scoring engine takes a
`MatchRules` as a parameter and behaves accordingly, so box cricket, gully
cricket, T20, and The Hundred are all the same engine with different config.
"""

from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field, field_validator, model_validator

from .enums import ALL_DISMISSALS, BallType, DismissalType


class PowerplayRange(BaseModel):
    """An inclusive, 1-based range of overs with a fielding restriction."""

    start_over: int = Field(ge=1)
    end_over: int = Field(ge=1)
    max_fielders_outside: int = Field(default=2, ge=0)
    label: str = "Powerplay"

    @model_validator(mode="after")
    def _check_range(self) -> "PowerplayRange":
        if self.end_over < self.start_over:
            raise ValueError("powerplay end_over must be >= start_over")
        return self

    def covers(self, over_number: int) -> bool:
        return self.start_over <= over_number <= self.end_over


class WideRules(BaseModel):
    enabled: bool = True
    run_penalty: int = Field(default=1, ge=0)
    counts_as_legal_ball: bool = False  # False => re-bowled (doesn't advance the over)
    allow_byes: bool = True  # can batters physically run extra on a wide?


class NoBallRules(BaseModel):
    enabled: bool = True
    run_penalty: int = Field(default=1, ge=0)
    counts_as_legal_ball: bool = False  # False => re-bowled
    free_hit: bool = True  # next legal delivery is a free hit
    off_bat_counts: bool = True  # runs hit off a no-ball go to the batter
    allow_byes: bool = True  # byes/leg-byes runnable after a no-ball


class Interruption(BaseModel):
    """One stoppage in play (rain, bad light, …) that may cut the overs.

    Lives on the match's own rulebook (persists in matches.rules JSON, no schema
    migration) and is re-applied on rebuild. Overs are *overs remaining* at the
    moment of the stoppage: the cut from ``overs_before`` → ``overs_after`` is what
    costs DLS resources. ``pending`` is True between the interruption and its resume.
    """

    innings: int = Field(ge=1, le=4)
    reason: str = "rain"              # rain | bad_light | wet_outfield | ground_delay | power_failure | other
    balls: int = 0                    # legal balls bowled in that innings when play stopped
    wickets: int = 0
    overs_before: float = 0.0         # overs remaining before the cut
    overs_after: float = 0.0          # overs remaining after resumption (≤ overs_before)
    interrupt_at: Optional[str] = None  # ISO time play stopped
    resume_at: Optional[str] = None     # ISO time play resumed
    pending: bool = False             # awaiting resume + overs confirmation

    @property
    def overs_lost(self) -> float:
        return round(max(0.0, self.overs_before - self.overs_after), 2)


class SuperOverRound(BaseModel):
    """Persisted setup for one super-over round.

    A super over is a one-over-per-side eliminator (all out at 2 wickets). Only
    the *setup* is stored here — who bats first, and whether the chase has begun;
    the deliveries themselves live in the event log as extra innings.
    """

    bat_first: str  # team name batting first this round
    second_started: bool = False


class MatchRules(BaseModel):
    """A complete, self-contained description of how a match is scored."""

    name: str = "Custom"
    format_id: str = "custom"
    description: str = ""

    # Shape of the contest
    players_per_side: int = Field(default=11, ge=2, le=20)
    overs_per_innings: int = Field(default=20, ge=1, le=200)
    balls_per_over: int = Field(default=6, ge=1, le=12)
    innings_per_side: int = Field(default=1, ge=1, le=2)
    ball_type: BallType = BallType.LEATHER

    # Bowling constraints
    max_overs_per_bowler: Optional[int] = Field(default=None, ge=1)
    allow_consecutive_overs: bool = False  # can a bowler bowl back-to-back overs?

    # Fielding restrictions
    powerplays: list[PowerplayRange] = Field(default_factory=list)
    # Max fielders allowed outside the 30-yard circle in normal (non-powerplay)
    # overs. Powerplay overs use their own (tighter) PowerplayRange limit.
    default_fielders_outside: int = Field(default=5, ge=0, le=20)

    # Extras
    wide: WideRules = Field(default_factory=WideRules)
    no_ball: NoBallRules = Field(default_factory=NoBallRules)
    byes_allowed: bool = True
    leg_byes_allowed: bool = True

    # Format quirks
    last_man_stands: bool = False  # last batter bats on alone (indoor/box)
    allow_declaration: bool = False
    super_over_on_tie: bool = False  # a tie is broken by a one-over eliminator
    # Rain/short-match revised targets (Duckworth–Lewis–Stern style). When on,
    # the match is flagged as DLS-eligible and the scorer may set a revised
    # target/overs; the automatic par-score table is not computed here.
    dls_enabled: bool = False
    # Per-match DLS overrides (NOT part of a saved template — they live on the
    # match's own rulebook so they persist in the matches.rules JSON with no
    # schema migration). Set by the scorer during a rain break; they reshape the
    # second innings (target to win + overs available).
    revised_target: Optional[int] = Field(default=None, ge=1)
    revised_overs: Optional[int] = Field(default=None, ge=1)
    # Rain interruptions (reason + times + overs cut). They DRIVE the automatic DLS
    # target — the scorer only confirms overs; revised_target/overs above are the
    # computed result. Persisted in rules JSON, re-applied on rebuild (no migration).
    interruptions: list["Interruption"] = Field(default_factory=list)
    # Match called off. abandoned=True with a completed 2nd innings past the DLS
    # minimum overs → decided on par; otherwise "No result".
    abandoned: bool = False
    abandon_reason: Optional[str] = None
    # Minimum overs the chasing side must face for a DLS result to stand (ICC: 20
    # for 50-over, 5 for T20). 0 = derive from the innings length.
    dls_min_overs: int = Field(default=0, ge=0)
    g50: int = Field(default=245, ge=1)  # avg 50-over score, for the "team 2 gained resources" case
    # Captain's declaration: which innings (1 or 2) was declared closed early.
    # Lives on the match's own rulebook (persists in matches.rules JSON, no schema
    # migration) and is re-applied on rebuild to keep that innings complete.
    declared_innings: Optional[int] = Field(default=None, ge=1, le=2)
    # Super-over rounds played to break a tie (one entry per round). Lives on the
    # match's own rulebook so the event log replays into the right innings with
    # no schema migration; empty for a saved template.
    super_over_rounds: list[SuperOverRound] = Field(default_factory=list)
    # "Rule-out" (vs "full ground"): hitting the ball over the boundary on the
    # full is OUT, not six — common in box/gully cricket with limited space.
    over_boundary_out: bool = False

    # Dismissals permitted in this format (e.g. gully cricket may ban LBW)
    allowed_dismissals: list[DismissalType] = Field(
        default_factory=lambda: list(ALL_DISMISSALS)
    )

    # Boundary values (kept configurable for exotic variants)
    four_value: int = 4
    six_value: int = 6

    @field_validator("max_overs_per_bowler")
    @classmethod
    def _cap_bowler_overs(cls, v: Optional[int], info) -> Optional[int]:
        return v

    @model_validator(mode="after")
    def _coherence(self) -> "MatchRules":
        if self.max_overs_per_bowler and self.max_overs_per_bowler > self.overs_per_innings:
            raise ValueError("max_overs_per_bowler cannot exceed overs_per_innings")
        for pp in self.powerplays:
            if pp.end_over > self.overs_per_innings:
                raise ValueError("powerplay range exceeds overs_per_innings")
        return self

    # ----- derived helpers used by the engine -----

    @property
    def wickets_to_all_out(self) -> int:
        """How many wickets end the innings.

        Normally players-1 (last batter has no partner). With last-man-stands the
        final batter continues alone, so it takes one more wicket to end it.
        """
        return self.players_per_side if self.last_man_stands else self.players_per_side - 1

    @property
    def total_legal_balls(self) -> int:
        return self.overs_per_innings * self.balls_per_over

    def is_powerplay(self, over_number: int) -> bool:
        return any(pp.covers(over_number) for pp in self.powerplays)

    def active_powerplay(self, over_number: int) -> Optional[PowerplayRange]:
        """The powerplay range covering this (1-based) over, if any."""
        for pp in self.powerplays:
            if pp.covers(over_number):
                return pp
        return None

    def fielders_outside_limit(self, over_number: int) -> int:
        """Max fielders allowed outside the circle in this (1-based) over.

        A powerplay's own restriction applies during its overs; otherwise the
        normal-over default governs.
        """
        pp = self.active_powerplay(over_number)
        return pp.max_fielders_outside if pp else self.default_fielders_outside

    def dismissal_allowed(self, d: DismissalType) -> bool:
        # "Boundary out" only exists in rule-out formats; everything else is
        # governed by the configured allow-list.
        if d is DismissalType.BOUNDARY_OUT:
            return self.over_boundary_out
        return d in self.allowed_dismissals

File: mobile/lib/core/models/rules.dart
/// Full-fidelity mirror of `backend/app/domain/rules.py :: MatchRules`.
///
/// Every field round-trips, so a rulebook fetched from `/presets/{id}` can be
/// edited in the rule builder and posted straight back inside `POST /matches`
/// without silently dropping anything the server cares about.
library;

import 'json.dart';

/// Every dismissal kind the engine understands.
class Dismissals {
  const Dismissals._();

  static const bowled = 'bowled';
  static const caught = 'caught';
  static const caughtBehind = 'caught_behind';
  static const caughtAndBowled = 'caught_and_bowled';
  static const lbw = 'lbw';
  static const runOut = 'run_out';
  static const stumped = 'stumped';
  static const hitWicket = 'hit_wicket';
  static const retiredHurt = 'retired_hurt';
  static const retiredOut = 'retired_out';
  static const obstructingField = 'obstructing_field';
  static const hitBallTwice = 'hit_ball_twice';
  static const timedOut = 'timed_out';
  static const boundaryOut = 'boundary_out';

  /// The default "everything allowed" set. `boundary_out` is excluded because it
  /// is format-specific, gated by [MatchRules.overBoundaryOut].
  static const standard = <String>[
    bowled,
    caught,
    caughtBehind,
    caughtAndBowled,
    lbw,
    runOut,
    stumped,
    hitWicket,
    retiredHurt,
    retiredOut,
    obstructingField,
    hitBallTwice,
    timedOut,
  ];

  /// On a free hit a batter can only go these ways.
  static const onFreeHit = <String>{runOut, obstructingField, hitBallTwice};

  /// Dismissals where naming the fielder makes sense.
  static const needsFielder = <String>{caught, caughtBehind, runOut, stumped};

  /// Dismissals where the batters can have completed runs before the ball was
  /// dead. Everywhere else the ball is dead at once, and the engine would still
  /// credit any runs sent with it to the striker — so the picker must not
  /// appear. Retirements are excluded because the striker who is credited need
  /// not be the batter leaving, and a double hit scores no runs off the bat.
  static const canCompleteRuns = <String>{runOut, obstructingField};

  /// Dismissals credited to the bowler.
  static const creditedToBowler = <String>{
    bowled,
    caught,
    caughtBehind,
    caughtAndBowled,
    lbw,
    stumped,
    hitWicket,
    boundaryOut,
  };

  /// Retired hurt leaves the batter not out.
  static bool isNotOut(String kind) => kind == retiredHurt;

  static const _labels = <String, String>{
    bowled: 'Bowled',
    caught: 'Caught',
    caughtBehind: 'Caught behind',
    caughtAndBowled: 'Caught & bowled',
    lbw: 'LBW',
    runOut: 'Run out',
    stumped: 'Stumped',
    hitWicket: 'Hit wicket',
    retiredHurt: 'Retired hurt',
    retiredOut: 'Retired out',
    obstructingField: 'Obstructing the field',
    hitBallTwice: 'Hit the ball twice',
    timedOut: 'Timed out',
    boundaryOut: 'Boundary out',
  };

  static String label(String kind) =>
      _labels[kind] ?? kind.replaceAll('_', ' ');
}

/// Physical ball used, which segments a player's career splits.
class BallTypes {
  const BallTypes._();
  static const leather = 'leather';
  static const tennis = 'tennis';
  static const other = 'other';
  static const all = <String>[leather, tennis, other];
  static String label(String v) => switch (v) {
        leather => 'Leather',
        tennis => 'Tennis',
        _ => 'Other',
      };
}

/// An inclusive, 1-based range of overs with a fielding restriction.
class PowerplayRange {
  final int startOver;
  final int endOver;
  final int maxFieldersOutside;
  final String label;

  const PowerplayRange({
    required this.startOver,
    required this.endOver,
    this.maxFieldersOutside = 2,
    this.label = 'Powerplay',
  });

  factory PowerplayRange.fromJson(Map<String, dynamic> j) => PowerplayRange(
        startOver: asInt(j['start_over'], 1),
        endOver: asInt(j['end_over'], 1),
        maxFieldersOutside: asInt(j['max_fielders_outside'], 2),
        label: asStr(j['label'], 'Powerplay'),
      );

  Map<String, dynamic> toJson() => {
        'start_over': startOver,
        'end_over': endOver,
        'max_fielders_outside': maxFieldersOutside,
        'label': label,
      };

  PowerplayRange copyWith({
    int? startOver,
    int? endOver,
    int? maxFieldersOutside,
    String? label,
  }) =>
      PowerplayRange(
        startOver: startOver ?? this.startOver,
        endOver: endOver ?? this.endOver,
        maxFieldersOutside: maxFieldersOutside ?? this.maxFieldersOutside,
        label: label ?? this.label,
      );

  bool covers(int overNumber) =>
      overNumber >= startOver && overNumber <= endOver;
}

class WideRules {
  final bool enabled;
  final int runPenalty;

  /// `false` means the delivery is re-bowled (it does not advance the over).
  final bool countsAsLegalBall;
  final bool allowByes;

  const WideRules({
    this.enabled = true,
    this.runPenalty = 1,
    this.countsAsLegalBall = false,
    this.allowByes = true,
  });

  factory WideRules.fromJson(Map<String, dynamic> j) => WideRules(
        enabled: asBool(j['enabled'], true),
        runPenalty: asInt(j['run_penalty'], 1),
        countsAsLegalBall: asBool(j['counts_as_legal_ball']),
        allowByes: asBool(j['allow_byes'], true),
      );

  Map<String, dynamic> toJson() => {
        'enabled': enabled,
        'run_penalty': runPenalty,
        'counts_as_legal_ball': countsAsLegalBall,
        'allow_byes': allowByes,
      };

  WideRules copyWith({
    bool? enabled,
    int? runPenalty,
    bool? countsAsLegalBall,
    bool? allowByes,
  }) =>
      WideRules(
        enabled: enabled ?? this.enabled,
        runPenalty: runPenalty ?? this.runPenalty,
        countsAsLegalBall: countsAsLegalBall ?? this.countsAsLegalBall,
        allowByes: allowByes ?? this.allowByes,
      );
}

class NoBallRules {
  final bool enabled;
  final int runPenalty;
  final bool countsAsLegalBall;

  /// The next legal delivery is a free hit.
  final bool freeHit;

  /// Runs hit off a no-ball are credited to the batter.
  final bool offBatCounts;
  final bool allowByes;

  const NoBallRules({
    this.enabled = true,
    this.runPenalty = 1,
    this.countsAsLegalBall = false,
    this.freeHit = true,
    this.offBatCounts = true,
    this.allowByes = true,
  });

  factory NoBallRules.fromJson(Map<String, dynamic> j) => NoBallRules(
        enabled: asBool(j['enabled'], true),
        runPenalty: asInt(j['run_penalty'], 1),
        countsAsLegalBall: asBool(j['counts_as_legal_ball']),
        freeHit: asBool(j['free_hit'], true),
        offBatCounts: asBool(j['off_bat_counts'], true),
        allowByes: asBool(j['allow_byes'], true),
      );

  Map<String, dynamic> toJson() => {
        'enabled': enabled,
        'run_penalty': runPenalty,
        'counts_as_legal_ball': countsAsLegalBall,
        'free_hit': freeHit,
        'off_bat_counts': offBatCounts,
        'allow_byes': allowByes,
      };

  NoBallRules copyWith({
    bool? enabled,
    int? runPenalty,
    bool? countsAsLegalBall,
    bool? freeHit,
    bool? offBatCounts,
    bool? allowByes,
  }) =>
      NoBallRules(
        enabled: enabled ?? this.enabled,
        runPenalty: runPenalty ?? this.runPenalty,
        countsAsLegalBall: countsAsLegalBall ?? this.countsAsLegalBall,
        freeHit: freeHit ?? this.freeHit,
        offBatCounts: offBatCounts ?? this.offBatCounts,
        allowByes: allowByes ?? this.allowByes,
      );
}

/// One stoppage in play. The server maintains these; the client only displays
/// them and echoes them back untouched.
class Interruption {
  final int innings;
  final String reason;
  final int balls;
  final int wickets;
  final double oversBefore;
  final double oversAfter;
  final String? interruptAt;
  final String? resumeAt;
  final bool pending;

  const Interruption({
    required this.innings,
    this.reason = 'rain',
    this.balls = 0,
    this.wickets = 0,
    this.oversBefore = 0,
    this.oversAfter = 0,
    this.interruptAt,
    this.resumeAt,
    this.pending = false,
  });

  factory Interruption.fromJson(Map<String, dynamic> j) => Interruption(
        innings: asInt(j['innings'], 1),
        reason: asStr(j['reason'], 'rain'),
        balls: asInt(j['balls']),
        wickets: asInt(j['wickets']),
        oversBefore: asDouble(j['overs_before']),
        oversAfter: asDouble(j['overs_after']),
        interruptAt: asStrOrNull(j['interrupt_at']),
        resumeAt: asStrOrNull(j['resume_at']),
        pending: asBool(j['pending']),
      );

  Map<String, dynamic> toJson() => {
        'innings': innings,
        'reason': reason,
        'balls': balls,
        'wickets': wickets,
        'overs_before': oversBefore,
        'overs_after': oversAfter,
        'interrupt_at': interruptAt,
        'resume_at': resumeAt,
        'pending': pending,
      };

  double get oversLost {
    final lost = oversBefore - oversAfter;
    return lost <= 0 ? 0 : double.parse(lost.toStringAsFixed(2));
  }
}

/// Reasons play can stop, matching the server's validation pattern.
class StoppageReasons {
  const StoppageReasons._();
  static const all = <String>[
    'rain',
    'bad_light',
    'wet_outfield',
    'ground_delay',
    'power_failure',
    'other',
  ];

  static String label(String v) => switch (v) {
        'rain' => 'Rain',
        'bad_light' => 'Bad light',
        'wet_outfield' => 'Wet outfield',
        'ground_delay' => 'Ground delay',
        'power_failure' => 'Power failure',
        _ => 'Other',
      };
}

/// Persisted setup for one super-over round.
class SuperOverRound {
  final String batFirst;
  final bool secondStarted;

  const SuperOverRound({required this.batFirst, this.secondStarted = false});

  factory SuperOverRound.fromJson(Map<String, dynamic> j) => SuperOverRound(
        batFirst: asStr(j['bat_first']),
        secondStarted: asBool(j['second_started']),
      );

  Map<String, dynamic> toJson() =>
      {'bat_first': batFirst, 'second_started': secondStarted};
}

/// A complete, self-contained description of how a match is scored.
class MatchRules {
  final String name;
  final String formatId;
  final String description;

  // ---- Shape of the contest ----
  final int playersPerSide;
  final int oversPerInnings;
  final int ballsPerOver;
  final int inningsPerSide;
  final String ballType;

  // ---- Bowling constraints ----
  final int? maxOversPerBowler;
  final bool allowConsecutiveOvers;

  // ---- Fielding restrictions ----
  final List<PowerplayRange> powerplays;
  final int defaultFieldersOutside;

  // ---- Extras ----
  final WideRules wide;
  final NoBallRules noBall;
  final bool byesAllowed;
  final bool legByesAllowed;

  // ---- Format quirks ----
  final bool lastManStands;
  final bool allowDeclaration;
  final bool superOverOnTie;
  final bool dlsEnabled;

  // ---- Per-match state the server owns (echoed back unchanged) ----
  final int? revisedTarget;
  final int? revisedOvers;
  final List<Interruption> interruptions;
  final bool abandoned;
  final String? abandonReason;
  final int dlsMinOvers;
  final int g50;
  final int? declaredInnings;
  final List<SuperOverRound> superOverRounds;

  /// Hitting the ball over the boundary on the full is OUT, not six.
  final bool overBoundaryOut;

  final List<String> allowedDismissals;
  final int fourValue;
  final int sixValue;

  const MatchRules({
    this.name = 'Custom',
    this.formatId = 'custom',
    this.description = '',
    this.playersPerSide = 11,
    this.oversPerInnings = 20,
    this.ballsPerOver = 6,
    this.inningsPerSide = 1,
    this.ballType = BallTypes.leather,
    this.maxOversPerBowler,
    this.allowConsecutiveOvers = false,
    this.powerplays = const [],
    this.defaultFieldersOutside = 5,
    this.wide = const WideRules(),
    this.noBall = const NoBallRules(),
    this.byesAllowed = true,
    this.legByesAllowed = true,
    this.lastManStands = false,
    this.allowDeclaration = false,
    this.superOverOnTie = false,
    this.dlsEnabled = false,
    this.revisedTarget,
    this.revisedOvers,
    this.interruptions = const [],
    this.abandoned = false,
    this.abandonReason,
    this.dlsMinOvers = 0,
    this.g50 = 245,
    this.declaredInnings,
    this.superOverRounds = const [],
    this.overBoundaryOut = false,
    this.allowedDismissals = Dismissals.standard,
    this.fourValue = 4,
    this.sixValue = 6,
  });

  factory MatchRules.fromJson(Map<String, dynamic> j) => MatchRules(
        name: asStr(j['name'], 'Custom'),
        formatId: asStr(j['format_id'], 'custom'),
        description: asStr(j['description']),
        playersPerSide: asInt(j['players_per_side'], 11),
        oversPerInnings: asInt(j['overs_per_innings'], 20),
        ballsPerOver: asInt(j['balls_per_over'], 6),
        inningsPerSide: asInt(j['innings_per_side'], 1),
        ballType: asStr(j['ball_type'], BallTypes.leather),
        maxOversPerBowler: asIntOrNull(j['max_overs_per_bowler']),
        allowConsecutiveOvers: asBool(j['allow_consecutive_overs']),
        powerplays:
            asMapList(j['powerplays']).map(PowerplayRange.fromJson).toList(),
        defaultFieldersOutside: asInt(j['default_fielders_outside'], 5),
        wide: WideRules.fromJson(asMap(j['wide'])),
        noBall: NoBallRules.fromJson(asMap(j['no_ball'])),
        byesAllowed: asBool(j['byes_allowed'], true),
        legByesAllowed: asBool(j['leg_byes_allowed'], true),
        lastManStands: asBool(j['last_man_stands']),
        allowDeclaration: asBool(j['allow_declaration']),
        superOverOnTie: asBool(j['super_over_on_tie']),
        dlsEnabled: asBool(j['dls_enabled']),
        revisedTarget: asIntOrNull(j['revised_target']),
        revisedOvers: asIntOrNull(j['revised_overs']),
        interruptions:
            asMapList(j['interruptions']).map(Interruption.fromJson).toList(),
        abandoned: asBool(j['abandoned']),
        abandonReason: asStrOrNull(j['abandon_reason']),
        dlsMinOvers: asInt(j['dls_min_overs']),
        g50: asInt(j['g50'], 245),
        declaredInnings: asIntOrNull(j['declared_innings']),
        superOverRounds: asMapList(j['super_over_rounds'])
            .map(SuperOverRound.fromJson)
            .toList(),
        overBoundaryOut: asBool(j['over_boundary_out']),
        allowedDismissals: j['allowed_dismissals'] is List
            ? asStrList(j['allowed_dismissals'])
            : Dismissals.standard,
        fourValue: asInt(j['four_value'], 4),
        sixValue: asInt(j['six_value'], 6),
      );

  Map<String, dynamic> toJson() => {
        'name': name,
        'format_id': formatId,
        'description': description,
        'players_per_side': playersPerSide,
        'overs_per_innings': oversPerInnings,
        'balls_per_over': ballsPerOver,
        'innings_per_side': inningsPerSide,
        'ball_type': ballType,
        'max_overs_per_bowler': maxOversPerBowler,
        'allow_consecutive_overs': allowConsecutiveOvers,
        'powerplays': powerplays.map((p) => p.toJson()).toList(),
        'default_fielders_outside': defaultFieldersOutside,
        'wide': wide.toJson(),
        'no_ball': noBall.toJson(),
        'byes_allowed': byesAllowed,
        'leg_byes_allowed': legByesAllowed,
        'last_man_stands': lastManStands,
        'allow_declaration': allowDeclaration,
        'super_over_on_tie': superOverOnTie,
        'dls_enabled': dlsEnabled,
        'revised_target': revisedTarget,
        'revised_overs': revisedOvers,
        'interruptions': interruptions.map((e) => e.toJson()).toList(),
        'abandoned': abandoned,
        'abandon_reason': abandonReason,
        'dls_min_overs': dlsMinOvers,
        'g50': g50,
        'declared_innings': declaredInnings,
        'super_over_rounds': superOverRounds.map((e) => e.toJson()).toList(),
        'over_boundary_out': overBoundaryOut,
        'allowed_dismissals': allowedDismissals,
        'four_value': fourValue,
        'six_value': sixValue,
      };

  MatchRules copyWith({
    String? name,
    String? formatId,
    String? description,
    int? playersPerSide,
    int? oversPerInnings,
    int? ballsPerOver,
    int? inningsPerSide,
    String? ballType,
    int? maxOversPerBowler,
    bool clearMaxOversPerBowler = false,
    bool? allowConsecutiveOvers,
    List<PowerplayRange>? powerplays,
    int? defaultFieldersOutside,
    WideRules? wide,
    NoBallRules? noBall,
    bool? byesAllowed,
    bool? legByesAllowed,
    bool? lastManStands,
    bool? allowDeclaration,
    bool? superOverOnTie,
    bool? dlsEnabled,
    int? dlsMinOvers,
    int? g50,
    bool? overBoundaryOut,
    List<String>? allowedDismissals,
    int? fourValue,
    int? sixValue,
  }) =>
      MatchRules(
        name: name ?? this.name,
        formatId: formatId ?? this.formatId,
        description: description ?? this.description,
        playersPerSide: playersPerSide ?? this.playersPerSide,
        oversPerInnings: oversPerInnings ?? this.oversPerInnings,
        ballsPerOver: ballsPerOver ?? this.ballsPerOver,
        inningsPerSide: inningsPerSide ?? this.inningsPerSide,
        ballType: ballType ?? this.ballType,
        maxOversPerBowler: clearMaxOversPerBowler
            ? null
            : (maxOversPerBowler ?? this.maxOversPerBowler),
        allowConsecutiveOvers:
            allowConsecutiveOvers ?? this.allowConsecutiveOvers,
        powerplays: powerplays ?? this.powerplays,
        defaultFieldersOutside:
            defaultFieldersOutside ?? this.defaultFieldersOutside,
        wide: wide ?? this.wide,
        noBall: noBall ?? this.noBall,
        byesAllowed: byesAllowed ?? this.byesAllowed,
        legByesAllowed: legByesAllowed ?? this.legByesAllowed,
        lastManStands: lastManStands ?? this.lastManStands,
        allowDeclaration: allowDeclaration ?? this.allowDeclaration,
        superOverOnTie: superOverOnTie ?? this.superOverOnTie,
        dlsEnabled: dlsEnabled ?? this.dlsEnabled,
        revisedTarget: revisedTarget,
        revisedOvers: revisedOvers,
        interruptions: interruptions,
        abandoned: abandoned,
        abandonReason: abandonReason,
        dlsMinOvers: dlsMinOvers ?? this.dlsMinOvers,
        g50: g50 ?? this.g50,
        declaredInnings: declaredInnings,
        superOverRounds: superOverRounds,
        overBoundaryOut: overBoundaryOut ?? this.overBoundaryOut,
        allowedDismissals: allowedDismissals ?? this.allowedDismissals,
        fourValue: fourValue ?? this.fourValue,
        sixValue: sixValue ?? this.sixValue,
      );

  /// How many wickets end the innings. Normally players-1 (the last batter has
  /// no partner); with last-man-stands the final batter continues alone.
  int get wicketsToAllOut =>
      lastManStands ? playersPerSide : playersPerSide - 1;

  int get totalLegalBalls => oversPerInnings * ballsPerOver;

  bool isPowerplay(int overNumber) =>
      powerplays.any((pp) => pp.covers(overNumber));

  PowerplayRange? activePowerplay(int overNumber) {
    for (final pp in powerplays) {
      if (pp.covers(overNumber)) return pp;
    }
    return null;
  }

  int fieldersOutsideLimit(int overNumber) =>
      activePowerplay(overNumber)?.maxFieldersOutside ?? defaultFieldersOutside;

  bool dismissalAllowed(String kind) => kind == Dismissals.boundaryOut
      ? overBoundaryOut
      : allowedDismissals.contains(kind);

  /// Every dismissal the scorer may pick in this format, in a sensible order.
  List<String> get selectableDismissals => [
        ...Dismissals.standard.where(allowedDismissals.contains),
        if (overBoundaryOut) Dismissals.boundaryOut,
      ];

  /// A one-line summary for cards and pickers, e.g. "20 ov · 11-a-side · Leather".
  String get summaryLine {
    final bits = <String>[
      '$oversPerInnings ov',
      '$playersPerSide-a-side',
      BallTypes.label(ballType),
      if (ballsPerOver != 6) '$ballsPerOver balls/over',
      if (lastManStands) 'Last man stands',
      if (overBoundaryOut) 'Rule-out',
    ];
    return bits.join(' · ');
  }
}

/// A built-in format as listed by `GET /presets`.
class PresetSummary {
  final String id;
  final String name;
  final String description;
  final int playersPerSide;
  final int oversPerInnings;
  final int ballsPerOver;
  final String ballType;
  final bool lastManStands;

  const PresetSummary({
    required this.id,
    required this.name,
    this.description = '',
    this.playersPerSide = 11,
    this.oversPerInnings = 20,
    this.ballsPerOver = 6,
    this.ballType = BallTypes.leather,
    this.lastManStands = false,
  });

  factory PresetSummary.fromJson(Map<String, dynamic> j) => PresetSummary(
        id: asStr(j['id']),
        name: asStr(j['name']),
        description: asStr(j['description']),
        playersPerSide: asInt(j['players_per_side'], 11),
        oversPerInnings: asInt(j['overs_per_innings'], 20),
        ballsPerOver: asInt(j['balls_per_over'], 6),
        ballType: asStr(j['ball_type'], BallTypes.leather),
        lastManStands: asBool(j['last_man_stands']),
      );

  String get summaryLine =>
      '$oversPerInnings ov · $playersPerSide-a-side · ${BallTypes.label(ballType)}';
}
